docs/git.rst

```diff
@@ -0,0 +1,304 @@
+
+This is specially important when trying to understand why some changes were made,
+or who did actually change that line, etc.
+
+Further explanation: http://stevenharman.net/git-pull-with-automatic-rebase
+
+Reviewing your changes
+----------------------
+After hacking for some minutes/hours you are finished and about to commit your changes,
+great!
+
+*BUT* (capital letters),
+please,
+do so with :command:`git add -p`.
+
+The ``-p`` switch allows you to select which hunks you want to add on a commit.
+
+This is not only great to split changes into different commits,
+but is also the time when you actually **review** your code before anyone else sees it.
+
+This is the time when you spot typos,
+pep8 errors,
+misaligned code,
+lack of docstrings in methods...
+
+Remember that the first code review is the one you do on your own.
+Some inspiration/better phrased:
+http://ada.mbecker.cc/2012/11/22/be-your-own-code-review/
+
+And please,
+do remember the gold metric about reviewing code:
+http://www.osnews.com/story/19266/WTFs_m
+
+One commit does one thing
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Repeat with me:
+*One commit does one thing*.
+Period.
+
+When someone else needs to review your code,
+most probably will give up or just skim over your code if that means that you just throw together a commit with +20 files doing all sorts of changes on them.
+
+If you can not express what has been changed within 50 characters
+(suggested length of a commit message subject),
+or you say it like "it does XXX and YYY",
+you most probably need to split that commit into,
+at least,
+two or three commits.
+
+That doesn't mean that a +20 files or +100 lines of code changes are bad per se,
+you may be doing a simple refactoring across lots of files,
+that's fine and good actually,
+as long as that commit is just and only about that refactoring,
+and not a refactoring plus moving things around plus fixing some bugs plus adding some docs plus a new cool feature plus fixing typos on documentation.
+
+And this is actually to help both your present self and your +5 years from now that will have to refactor that code of yours,
+and maybe is struggling to understand what was going on there.
+
+On top of that,
+small self-contained commits with descriptive enough commit messages,
+help keeping the `cognitive load <http://en.wikipedia.org/wiki/Cognitive_load>`_ down.
+
+That will make your changes easy to be reviewed by others,
+which in turn make it easier to accept or give feedback to your changes.
+
+Making commits
+--------------
+For commit messages see:
+`plone API guidelines<http://docs.plone.org/external/plone.api/docs/contribute/conventions.html#git-commit-message-style>`_.
+
+Adding references to issues
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Always add the full URL to the issue/pull request you are fixing/referring to.
+
+Maybe within the git repository it makes sense,
+but as soon as you are outside of it, it will not.
```

Comment:
Most probably, as I said I wanted to get a first version out and later review it.

In this section I was referring to the fact that if your commit message says:

    Fix issue #74

You can relate that issue 74 to the current project where the commit was pushed, but alas, p.a.discussion (for the sake of an example) does not have issues enabled on github, so where the hell is this issue 74 from?

My suggestion is to instead write it like:

    Fix issue about encoding
    
    La la la la la la la
    la la la la li li li li li
    
    Fixes issue: https://github.com/plone/Products.CMFPlone/issue/74